staging/src/k8s.io/apimachinery/pkg/util/validation/field/testing/match.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"testing"
+
+	field "k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// MatchErrors compares two ErrorLists with the specified matcher, and fails
+// the test if they don't match. If a given "want" error matches multiple "got"
+// errors, they will all be consumed. This might be OK (e.g. if there are
+// multiple errors on the same field from the same origin) or it might be an
+// insufficiently specific matcher, so these will be logged.
+func MatchErrors(t *testing.T, want, got field.ErrorList, matcher *Matcher) {
+	t.Helper()
+
+	remaining := got
+	for _, w := range want {
+		tmp := make(field.ErrorList, 0, len(remaining))
+		n := 0
+		for _, g := range remaining {
+			if matcher.Matches(w, g) {
+				n++
+			} else {
+				tmp = append(tmp, g)
+			}
+		}
+		if n == 0 {
+			t.Errorf("expected an error matching:\n%s", matcher.Render(w))
+		} else if n > 1 {
+			// This is not necessarily and error, but it's worth logging in
+			// case it's not what the test author intended.
+			t.Logf("multiple errors matched:\n%s", matcher.Render(w))
```

Comment:
Trying to catch up here.  Bare with me as I run some cases through my head:

- Migrate 2 imperative validations -> 1 declarative validation:  Use same origin for both imperative and declarative.  Tests pass.  Imperative validations log this message but we know it's fine and ignore it.
- Migrate 1 imperative validation -> 2 declarative validations: Use same origin for both imperative and declarative. Tests pass. Declarative validations log this message but we know it's fine and ignore it.
- Defect is introduced validation logic that produces duplicate validations:  Message is logged, but if the test is pre-existing, PR that introduces the defect is not the PR that adds the tests, this might not be caught?

Is the last one a problem or did I miss something?

Reply:
The last one is why I added the log.  I think we need to allow multiple errors of the same origin to by matched by a single match, decoupling tests from implementation.  But if you accidentally add the wrong origin to a new validation, I would hope that a test which should fail (we all write our tests to prove they fail, right?) won't fail, and MAYBE that dev will run the test with `-v` and see this and realize what they did.

Open to ideas how to do better, of course.

Reply:
Ok, I think this is fine for now.

Reply:
Late to the party, can be a follow-up: I think consuming multiple "got" errors with a single "want" error should by default be a test failure. It seems safer to me because the test is more precise and matches (no pun intended) my expectation how list comparisons work.

As it stands now, a developer might expect exactly one error. It passes when  getting. A developer might not notice while writing the test, or it is a regression later where no-one looks for log output.

If a developer wants to consume multiple errors, then this could be an opt-in flag in the matcher. More flexible would be a flag in the "want" list entry ("this error exactly once, this one potentially multiple times").

Reply:
Part of the purpose of this whole "origin" idea was to allow validation code to evolve from returning one error to returning multiple, and unless the test ACTUALLY CARES about how many (almost always reserved for the validation code's unit tests), it should not express that.  It's an  implementation detail and makes the tests brittle.

If your test cares to distinguish, it can use the `ByDetail*` settings.  I have yet to see a place where there are two errors on the same field, which are identical in Type, Field, Value, AND Details.

Reply:
> Part of the purpose of this whole "origin" idea was to allow validation code to evolve from returning one error to returning multiple, unless the test ACTUALLY CARES about how many (almost always reserved for the validation code's unit tests)

So the goal is to be fuzzy about errors "inherited" from validation code that is reused (= called indirectly). That makes sense to me. But then I'd still would like to be strict about errors that originate in the validation code that is being tested directly.

> If your test cares to distinguish, it can use the ByDetail* settings.

I'm not sure whether that works. I then also have to provide the details for "inherited" errors, which defeats the purpose of being more flexible about those.

Reply:
`ByDetailSubstring()` allows you to say `""` for those errors you do not care about but express unique substrings for those errors you do care about.

Reply:
I see - I had missed that particular option. That should indeed work.